### ISSUE TITLE
Authentication issues with TFS

### ISSUE THREAD

**Issue by RobCannon** [microsoft/vsts-agent-docker#55]
I was not able to get these images to work successfully with our on-premise TFS server, but I have a solution.

Our server is not using https, so trying to authenticate with a PAT token will fail.  The solution is the use Negotiate authentication.  Here is a part of how I modified the start.sh file in my container image:

./bin/Agent.Listener configure --unattended \
  --agent "${VSTS_AGENT:-$(hostname)}" \
  --url "$TFS_URL" \
  --auth Negotiate \
  --username "$TFS_USERNAME" \
  --password "$TFS_PASSWORD" \
  --pool "${VSTS_POOL:-Default}" \
  --work "${VSTS_WORK:-_work}" \
  --replace & wait $!

I would make a pull request to make this change, but I am unsure of some things.  This looks like it needs to be made in many files, and I am not sure which ones should be changed.  Also, I think the proper solution should support either using the VSTS_TOKEN variable or the TFS_USERNAME/TFS_PASSWORD combo.  My solution only works with the TFS_USERNAME/TFS_PASSWORD.  Finally, I wasn't sure of the best way to support credentials in a file for this scenario. 

Should I attempt to make a PR for this or let you guys make these changes?

**Comment by stepro** [comment]
Did you determine this actually works? I don't believe the Negotiate auth model works on Linux because it's for Windows authentication via NTLM or Kerberos.

**Comment by chechunawer** [comment]
I have finish an implementation for this login and works properly.
One thing I need to know, is how to clean pool after this docker has received halt signal.

**Comment by stepro** [comment]
The startup script should be configured to correctly remove the agent from the agent pool if you gracefully terminate the container. What do you see that isn't working?

**Comment by chechunawer** [comment]
With changes I have made in TFS there are agents stopped.
But it does not delete it.

**Comment by stepro** [comment]
If you terminate the container using SIGINT or SIGTERM, it will run [this](https://github.com/Microsoft/vsts-agent-docker/blob/master/ubuntu/tfs/start.sh#L34) code which should remove it from the pool. What does your implementation do? Maybe it conflicts with this startup script somehow?

**Comment by chechunawer** [comment]
Thank you,
I was not cleaning up correctly.
Now it clears correctly agents created.

**Comment by chechunawer** [comment]
[modifiedStartForNTLM.zip](https://github.com/Microsoft/vsts-agent-docker/files/1682625/modifiedStartForNTLM.zip)

If someone needs it, here is the file

**Comment by stepro** [comment]
Thanks. When I get time, I'll update the official start.sh script to support both negotiate and token based auth models.

**Comment by gravufo** [comment]
I think you guys should be using gMSA for authentication instead of passing username/password.
Not only would it be safer, but it's also easier if you want to access network resources afterwards (for example if you need to publish artifacts to a network share).

It goes like this:
1. Create gMSA in AD.
2. Authorize the docker host to access/use that gMSA in AD as well as on the machine itself.
3. Create a credentialspec for the user on the docker host machine.
4. Start your container using the credentialspec we created above.
5. Use --auth Integrated during agent configuration.

[Here](https://blogs.msdn.microsoft.com/containerstuff/2017/01/30/create-a-container-with-active-directory-support/) is the more doc on how to use gMSA.

Hope this helps!

**Comment by stepro** [comment]
@gravufo, will that work for Linux containers? The link shows a whole lot of Windows specific things.

**Comment by gravufo** [comment]
You're right, it won't. I hadn't though about connecting to an on-premise TFS from a linux container!
On the other hand, I'd still suggest supporting both ways, in the sense that if I do want to use NTLM and I'm on a Windows container, it would be nice to be able to take advantage of the gMSA and, of course, it would be nice to be able to use NTLM in a linux container using provided credentials.

So in total, you'd have three ways of authenticating:
1. PAT
2. NTLM using credentials
3. NTLM using gMSA

Thanks for the clarification :)

**Comment by stepro** [comment]
Yeah, the work here for the Linux container is to parameterize the auth model which is currently hardcoded to pat only.

**Comment by mjwoodsphx** [comment]
I have attempted to follow chechunqwer's instructions and createed a new image for docker as we also don't have TFS enalbed with SSL at the moment.  However, no matter what I do I get the dreaded "exec format error".  I have researched this to death and all paths lead to windows formatting being used in the files.  I have converted them to Linux format and still get this error. 

Will it be possible to merge chechunqwer's changes into the main project soon?  It would really help me out!
